core/types/header_ext_test.go

```diff
@@ -0,0 +1,244 @@
+// (c) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package types
+
+import (
+	"bytes"
+	"math/big"
+	"testing"
+
+	"github.com/ava-labs/libevm/common"
+	ethtypes "github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/rlp"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHeaderExtraGetSet(t *testing.T) {
+	t.Parallel()
+
+	h := &Header{}
+
+	extra := GetHeaderExtra(h)
+	require.NotNil(t, extra)
+	assert.Equal(t, &HeaderExtra{}, extra)
+
+	extra = &HeaderExtra{
+		ExtDataHash: [32]byte{1},
+	}
+	SetHeaderExtra(h, extra)
+
+	extra = GetHeaderExtra(h)
+	assert.Equal(t, &HeaderExtra{
+		ExtDataHash: [32]byte{1},
+	}, extra)
+}
+
+func TestHeaderExtraRLP(t *testing.T) {
+	t.Parallel()
+
+	eth := &ethtypes.Header{
+		ParentHash: common.Hash{1},
+	}
+	extra := &HeaderExtra{
+		ExtDataHash: common.Hash{2},
+	}
+
+	writer := bytes.NewBuffer(nil)
+	err := extra.EncodeRLP(eth, writer)
+	require.NoError(t, err)
+
+	stream := rlp.NewStream(bytes.NewReader(writer.Bytes()), 0)
+	decodedExtra := new(HeaderExtra)
+	decodedEth := new(ethtypes.Header)
+	err = decodedExtra.DecodeRLP(decodedEth, stream)
+	require.NoError(t, err)
+
+	wantEth := &ethtypes.Header{
+		ParentHash: common.Hash{1},
+		Difficulty: new(big.Int),
+		Number:     new(big.Int),
+		Extra:      []byte{},
+	}
+	assert.Equal(t, wantEth, decodedEth)
+
+	wantExtra := &HeaderExtra{
+		ExtDataHash: common.Hash{2},
+	}
+	assert.Equal(t, wantExtra, decodedExtra)
+}
+
+func TestHeaderExtraJSON(t *testing.T) {
+	t.Parallel()
+
+	eth := &ethtypes.Header{
+		ParentHash: common.Hash{1},
+		// Required json fields
+		Difficulty: big.NewInt(2),
+		Number:     big.NewInt(3),
+	}
+	extra := &HeaderExtra{
+		ExtDataHash: common.Hash{2},
+	}
+
+	encoded, err := extra.EncodeJSON(eth)
+	require.NoError(t, err)
+
+	decodedExtra := new(HeaderExtra)
+	decodedEth := new(ethtypes.Header)
+	err = decodedExtra.DecodeJSON(decodedEth, encoded)
+	require.NoError(t, err)
+
+	wantEth := &ethtypes.Header{
+		ParentHash: common.Hash{1},
+		Difficulty: big.NewInt(2),
+		Number:     big.NewInt(3),
+		Extra:      []byte{},
+	}
+	assert.Equal(t, wantEth, decodedEth)
+
+	wantExtra := &HeaderExtra{
+		ExtDataHash: common.Hash{2},
+	}
+	assert.Equal(t, wantExtra, decodedExtra)
+}
+
+func TestHeaderSerializable_updates(t *testing.T) {
```

Comment:
(subjective) I'm not a big fan of tests that target implementation details instead of the end goal, which is RLP and JSON round-tripping. I think these should be collapsed into one of the above tests, but that's just my opinion.

Rationale: if the implementation changes (which I strongly believe it should, in a move to `rlp.Fields` like with `Body` instead of a complete override) then this test becomes obsolete and the change is more dangerous. If the test is limited to the actual code path of interest then the test can remain unmodified and therefore provide far greater confidence. Effectively, refocusing "test behaviour, not implementation".

Reply:
No longer used with newer code, resolving this.